client/client.go

```diff
@@ -87,6 +114,12 @@ func NewConnected(conn net.Conn) (client *Client) {
 	return
 }
 
+func (client *Client) getConn() *connection {
+	client.Lock()
```

Comment:
Please do not add this lock, a lot of effort was put into removing the weird global locks in this code and replacing it with compare and swap. Your changes need to be resilient to that design decision because otherwise the lock (unless you hang onto it for the _duration of interacting with the connection_) is not actually doing you any good here.

Reply:
+1 to scott, this doesn't seem to do anything, since returning client.conn is just passing a pointer back.

Reply:
The purpose of this lock is actually to prevent the readLoop from using a connection while the writeLoop has seen an error on that connection and already started reconnect.